### PR TITLE
Implement named route helpers

### PR DESCRIPTION
**Pull request by reprah**
Hi,

This is a response to issue #863; it adds path helper functions to Grape projects. I also have a [fork of grape-on-rack](https://github.com/reprah/grape-on-rack/tree/named-route-helpers#get-redirected-grape-route-helpers-demonstration) to demonstrate it (there are new instructions in its README). If there's anything I can provide to make it more obvious how the feature works, or if I need to explain why I did something a certain way, please let me know.

Summary: Grape::Route objects have dynamically named methods that return their paths, and there's a module called NamedRouteMatcher that does nothing but use its `method_missing` to match a method name to those routes and returns a path. The module's included in Grape::Endpoint and can be included by itself into another project.

I tried to provide thorough test coverage, and all new files have been Rubocop'd.

I'd like your thoughts on two things in particular:

- The difference between including the route helpers module at the Grape::Endpoint class level vs. [extending the individual Grape::Endpoint instances](https://github.com/intridea/grape/blob/master/lib/grape/endpoint.rb#L188)
- When it's appropriate to make something a private method in this project

Thanks for any feedback.

### CODE REVIEW

**Comment by dblock** [comment]
First, this is some excellent work, thank you and very well done. The code is clean, this could be merged as is (I will do a more thorough review, but generally it looks very, very good).

We've discussed continuing a path of modularizing Grape. I wonder whether this could be implemented in a separate gem (eg. `grape-route-helpers`). What do you think?

**Comment by kuraga** [comment]
Excellent! Changelog, readme?

**Comment by dblock** [comment]
+1 @kuraga

File: lib/grape/api.rb [hunk 1]
```diff
@@ -50,12 +50,16 @@ def cascade(value = nil)
...
+        subclasses.flat_map(&:prepare_routes)
+      end
 
       def prepare_routes
         endpoints.map(&:routes).flatten
       end
 
+      protected
+
       # Execute first the provided block, then each of the
```
Comment (dm1try) [inline lib/grape/api.rb]: why `.prepare_routes" scope visibility was changed to `public`?

File: lib/grape/named_route_matcher.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+module Grape
+  module NamedRouteMatcher
+    def method_missing(method_id, *arguments)
+      segments = arguments.first || {}
+
+      route = Grape::API.all_routes.detect do |r|
```
Comment (dm1try) [inline lib/grape/named_route_matcher.rb]: Can be `Grape::API.all_routes` is memoized? or they change in the runtime?

File: lib/grape/named_route_matcher.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        route_match?(r, method_id, segments)
```
Comment (dm1try) [inline lib/grape/named_route_matcher.rb]: Could this `method_missing` use the "helper name" matching before do its job? It's too greedy for me. Seems you already have this method https://github.com/intridea/grape/pull/983/files#diff-64d352c6d46153e1e4ff964b0ff4ecdeR46 , maybe this matching can be done using this info?

**Comment by dm1try** [comment]
@reprah  good job!

+1  to the implementation in a separate gem

**Comment by dm1try** [comment]
one more question, we have this spec (".prepare_routes should be called once")- https://github.com/intridea/grape/blob/489cdc94a4e36a40f6707d55c565232336aae4fc/spec/grape/dsl/routing_spec.rb#L212
I mean, could be any side effects when you are using `.prepare_routes` directly instead of call `.routes`?

**Comment by reprah** [comment]
@dblock @dm1try Thanks the brief reviews and recognizing the effort I put into this. The way you all respond to pull requests and users in general is a big reason why I want to write things for Grape in the first place (apart from the fact I use it regularly).

I do remember dblock saying before that Grape was trying to become more modular. This can be its own gem - I can only think of two downsides:

1.) From a code standpoint, it would require reopening class definitions. It was also useful to have the entire Grape test suite uncover edge cases in the code I added.

2.) From a community standpoint, you might get duplicate requests from people wanting named route helpers (if they don't realize/don't look for a gem that does this).

I'd be happy with having my own gem and incorporating advice from code reviews here. If I add to Grape's README or wiki and link my gem, maybe that can prevent some of number 2. Thoughts?

**Comment by dblock** [comment]
Thanks for the kind words. Make a new gem, PR a link to the README. We should be treating such gems as *supported extensions*, so this should be one. At some point you should PR some very minimal tests into Grape proper that check that the extensions still work with the last shipped release of the extension.

Anything that cannot be done via a simple `extend Module` should be refactored in Grape to allow that.

**Comment by reprah** [comment]
Is there another gem that's an "extension" to Grape that I can use as a reference? Does the word extension have any special meaning here (i.e. does it mean the gem interacts with the grape gem in a specific way)?

**Comment by dblock** [comment]
@reprah I don't think so, you're pioneering something new!

**Comment by dblock** [comment]
@reprah Did you get a change to move forward with this?

**Comment by reprah** [comment]
@dblock I'm happily working on it! Thanks for checking in. I found some patterns I like in other grape-related gems, and I have a good idea of what kinds of tests would verify my gem still works with the current version of Grape.

**Comment by dblock** [comment]
Bump, someone attempted something similar in https://github.com/intridea/grape/pull/1040, maybe we should converge on an implementation?

**Comment by reprah** [comment]
How could we approach that?

I do like how their implementation allows custom method names.

**Comment by dblock** [comment]
I think whoever releases a gem first wins :) Then we can iterate on it.

**Comment by reprah** [comment]
Ha! Sounds good.

**Comment by reprah** [comment]
I pushed my gem today, repository is here: https://github.com/reprah/grape-route-helpers
I did experience some regressions with testing with 0.12.0, so I think your idea of adding tests to Grape that test that my gem works with it is probably a good idea. I'd only need to verify that some of the attributes on Grape::Route objects have the expected format, I think.

**Comment by dblock** [comment]
Cool. Open a PR with a 1-liner about named routes linking the gem from the README. It's a frequent-enough request that we should make it easily discoverable. 

I'll close this.

**Comment by dblock** [comment]
Also add it to the WIKI.
